### PR TITLE
Fix/console globals

### PR DESCRIPTION
**Pull request by cds-amal**
This PR
  - changes how the REPL context is instantiated and. 
  - removes annoying `undefined` from REPL output. (See [docs](https://nodejs.org/dist/latest-v14.x/docs/api/repl.html#repl_repl_start_options))
  - adds tests using the new framework @eggplantzzz introduced.
  - should resolve 
    - #4238 
    - #4239 

This is necessary to capture the globals that the REPL includes in the context, which needed for a good user experience. Being able to `console.log` contributes to good experience.

The motivation for the previous change was to allow the truffle context variables to be hydrated before processing user input which is needed for our tests.  This PR still allows for that by coordinating when to send the input from tests. The sequence of events is as follows:

Child Process:
  1. REPL starts with nodes context and prompt = ''
  1. REPL loads truffle specific variables into context
  1. REPL sets the prompt to '(develop)>'

Test runner:
  1. Scans child's sdtout looking for '(develop)>' before sending input

### CODE REVIEW

File: packages/core/lib/console.js [hunk 1]
```diff
@@ -61,22 +61,31 @@ class Console extends EventEmitter {
...
-      this.repl = { context };
-      // this hydrates the environment with the user's contracts
-      this.provision();
-
+      // start the repl with an empty prompt and show a proper one when
+      // the repl has set up its context and is ready to accept input
       this.repl = repl.start({
-        prompt: "truffle(" + this.options.network + ")> ",
+        prompt: "",
+        ignoreUndefined: true,
```
Comment (davidmurdoch) [inline packages/core/lib/console.js:68]: Is this change considered breaking?

Comment (cds-amal) [reply packages/core/lib/console.js:68]: I'm not sure, but leaning towards removing it.

Comment (cds-amal) [reply packages/core/lib/console.js:68]: [Adding documentation for reviewers]
[ignoreUndefined](https://nodejs.org/dist/latest-v14.x/docs/api/repl.html#repl_repl_start_options) :- <boolean> If true, specifies that the default writer will not output the return value of a command if it evaluates to undefined. **Default**: false.

File: packages/core/lib/console.js [hunk 2]
```diff
@@ -61,22 +61,31 @@ class Console extends EventEmitter {
...
 
-      //want repl to exit when it receives an exit command
+      // Get and set Truffle and User Globals
+      const truffleAndUserGlobals = await this.calculateTruffleAndUserGlobals() ;
+      Object.entries(truffleAndUserGlobals).forEach(([key, value]) => {
+        Object.defineProperty(this.repl.context, key, {
+          configurable: false,
+          enumerable: true,
+          value,
+        });
```
Comment (cds-amal) [inline packages/core/lib/console.js:79]: This is a breaking change that we may not want. Users will no longer be able to delete truffleAndUserGlobals from the global context. This means, accounts for example:
```
delete this.accounts // will not delete or reassign accounts
accounts.splice(1,5) // will still change the contents, along with push, pop
```

**Review by cds-amal — COMMENTED** [review]
wanted to point out a breaking change for discussion/review. I don't have a strong opinion on the behavior.

File: packages/core/lib/console.js (hunk 1 above)
Comment (eggplantzzz) [reply packages/core/lib/console.js:68]: Do we want this? It should behave like a normal Node repl, right?

File: packages/truffle/test/scenarios/commandRunner.js [hunk 3]
```diff
@@ -51,6 +54,20 @@ module.exports = {
       });
 
       child.stdout.on("data", data => {
+        // accumulate buffer from chunks
```
Comment (eggplantzzz) [inline packages/truffle/test/scenarios/commandRunner.js:57]: Is this not currently working the way we thought it was?

File: packages/core/lib/console.js [hunk 4]
```diff
@@ -61,22 +61,31 @@ class Console extends EventEmitter {
...
+        ignoreUndefined: true,
         eval: this.interpret.bind(this)
       });
-      this.repl.context = context;
 
-      //want repl to exit when it receives an exit command
+      // Get and set Truffle and User Globals
+      const truffleAndUserGlobals = await this.calculateTruffleAndUserGlobals() ;
+      Object.entries(truffleAndUserGlobals).forEach(([key, value]) => {
+        Object.defineProperty(this.repl.context, key, {
```
Comment (eggplantzzz) [inline packages/core/lib/console.js:75]: Is it important that we define these using `defineProperty`? Why couldn't we just push them more naïvely into the context like we were doing before?

File: packages/core/lib/console.js (hunk 1 above)
Comment (cds-amal) [reply packages/core/lib/console.js]: Removing it from this PR to be revisited in follow on work

File: packages/core/lib/console.js (hunk 2 above)
Comment (cds-amal) [reply packages/core/lib/console.js]: After discussing with @eggplantzzz we feel users should have unfettered access to the context.

File: packages/truffle/test/scenarios/commandRunner.js (hunk 3 above)
Comment (cds-amal) [reply packages/truffle/test/scenarios/commandRunner.js:57]: The child process will only print the repl prompt when it is ready (has its context loaded with truffle variables). The command runner monitors the child's output, looking for that repl prompt (flag) before sending data to it

File: packages/core/lib/console.js (hunk 4 above)
Comment (cds-amal) [reply packages/core/lib/console.js]: Reverted to prior behavior.

File: packages/core/lib/console.js [hunk 5]
```diff
@@ -61,22 +61,26 @@ class Console extends EventEmitter {
...
       this.repl = repl.start({
-        prompt: "truffle(" + this.options.network + ")> ",
+        prompt: "",
         eval: this.interpret.bind(this)
       });
-      this.repl.context = context;
 
-      //want repl to exit when it receives an exit command
+      // Get and set Truffle and User Globals
+      const truffleAndUserGlobals = await this.calculateTruffleAndUserGlobals() ;
```
Comment (eggplantzzz) [inline packages/core/lib/console.js:72]: There is an extra, unnecessary space at the end of the line

**Review by eggplantzzz — APPROVED** [review]
I think it looks good, I did a simple test and all the globals are there.

**Pull request merged**